### PR TITLE
add a new module to configure isis on interface.

### PR DESCRIPTION
**Pull request by yuandongx**
##### ISSUE TYPE

- New Module Pull Request

### test case
```
  - name: merge isis circuit-level 1
    ce_is_is_interface:
      instance_id: 3
      ifname: Eth-Trunk20
      leveltype: level_1
      level1dispriority: 10
      level1cost: 50

```
### log info
```

changed: [10.190.121.125] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "end_state": {
        "instance": {
            "circuitLevelType": "level_1",
            "ifName": "Eth-Trunk20",
            "level1Cost": "50",
            "level1DisPriority": "10",
            "level2Cost": null,
            "level2DisPriority": "64",
            "p2pNegotiationMode": null,
            "p2pPeerIPIgnore": "false",
            "pPPOsicpCheckEnable": "false",
            "silentCost": "false",
            "silentEnable": "false",
            "snpaCheck": "false",
            "typeP2pEnable": "false"
        }
    },
    "existing": {
        "instance": {}
    },
    "invocation": {
        "module_args": {
            "bfdblocken": null,
            "bfdstaticen": null,
            "host": "10.190.121.125",
            "ifname": "Eth-Trunk20",
            "instance_id": 3,
            "level1cost": 50,
            "level1dispriority": 10,
            "level2cost": null,
            "level2dispriority": null,
            "leveltype": "level_1",
            "p2pnegotiationmode": null,
            "p2ppeeripignore": null,
            "ppposicpcheckenable": null,
            "silentcost": null,
            "silentenable": null,
            "snpacheck": null,
            "state": "present",
            "typep2penable": null
        }
    },
    "proposed": {
        "bfdblocken": null,
        "bfdstaticen": null,
        "ifname": "Eth-Trunk20",
        "instance_id": 3,
        "level1cost": 50,
        "level1dispriority": 10,
        "level2cost": null,
        "level2dispriority": null,
        "leveltype": "level_1",
        "p2pnegotiationmode": null,
        "p2ppeeripignore": null,
        "ppposicpcheckenable": null,
        "silentcost": null,
        "silentenable": null,
        "snpacheck": null,
        "state": "present",
        "typep2penable": null
    },
    "updates": [
        "interface Eth-Trunk20",
        "isis enable 3",
        "isis circuit-level level-1",
        "isis dis-priority 10 level-1",
        "isis cost 50 level-1"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show on device
```
[~8850_130.12-Eth-Trunk20]di th
#
interface Eth-Trunk20
 undo portswitch
 ip address 30.2.2.6 255.255.255.0
 isis enable 3
 isis circuit-level level-1
 isis cost 50 level-1
 isis dis-priority 10 level-1
#
return
[~8850_130.12-Eth-Trunk20]

```
### case
```

  - name: merge isis circuit-level 2
    ce_is_is_interface:
      instance_id: 3
      ifname: Eth-Trunk20
      leveltype: level_2
      level2dispriority: 8
      level2cost: 40
      state: present

```
### log info
```

changed: [10.190.121.125] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "end_state": {
        "instance": {
            "circuitLevelType": "level_2",
            "ifName": "Eth-Trunk20",
            "level1Cost": "50",
            "level1DisPriority": "10",
            "level2Cost": "40",
            "level2DisPriority": "8",
            "p2pNegotiationMode": null,
            "p2pPeerIPIgnore": "false",
            "pPPOsicpCheckEnable": "false",
            "silentCost": "false",
            "silentEnable": "false",
            "snpaCheck": "false",
            "typeP2pEnable": "false"
        }
    },
    "existing": {
        "instance": {
            "circuitLevelType": "level_1",
            "ifName": "Eth-Trunk20",
            "level1Cost": "50",
            "level1DisPriority": "10",
            "level2Cost": null,
            "level2DisPriority": "64",
            "p2pNegotiationMode": null,
            "p2pPeerIPIgnore": "false",
            "pPPOsicpCheckEnable": "false",
            "silentCost": "false",
            "silentEnable": "false",
            "snpaCheck": "false",
            "typeP2pEnable": "false"
        }
    },
    "invocation": {
        "module_args": {
            "bfdblocken": null,
            "bfdstaticen": null,
            "host": "10.190.121.125",
            "ifname": "Eth-Trunk20",
            "instance_id": 3,
            "level1cost": null,
            "level1dispriority": null,
            "level2cost": 40,
            "level2dispriority": 8,
            "leveltype": "level_2",
            "p2pnegotiationmode": null,
            "p2ppeeripignore": null,
            "ppposicpcheckenable": null,
            "silentcost": null,
            "silentenable": null,
            "snpacheck": null,
            "state": "present",
            "typep2penable": null
        }
    },
    "proposed": {
        "bfdblocken": null,
        "bfdstaticen": null,
        "ifname": "Eth-Trunk20",
        "instance_id": 3,
        "level1cost": null,
        "level1dispriority": null,
        "level2cost": 40,
        "level2dispriority": 8,
        "leveltype": "level_2",
        "p2pnegotiationmode": null,
        "p2ppeeripignore": null,
        "ppposicpcheckenable": null,
        "silentcost": null,
        "silentenable": null,
        "snpacheck": null,
        "state": "present",
        "typep2penable": null
    },
    "updates": [
        "interface Eth-Trunk20",
        "isis enable 3",
        "isis circuit-level level-2",
        "isis dis-priority 8 level-2",
        "isis cost 40 level-2"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
### show on device
```
[~8850_130.12-Eth-Trunk20]di th
#
interface Eth-Trunk20
 undo portswitch
 ip address 30.2.2.6 255.255.255.0
 isis enable 3
 isis circuit-level level-2
 isis cost 50 level-1
 isis cost 40 level-2
 isis dis-priority 10 level-1
 isis dis-priority 8 level-2
#
return
[~8850_130.12-Eth-Trunk20]

```
### test case
```

  - name: merge isis typep2penable
    ce_is_is_interface:
      instance_id: 3
      ifname: Eth-Trunk20      
      typep2penable: true
      p2pnegotiationmode: 2_way
      p2ppeeripignore: true
      ppposicpcheckenable: true
      state: present

```
### log info
```

changed: [10.190.121.125] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "end_state": {
        "instance": {
            "circuitLevelType": "level_2",
            "ifName": "Eth-Trunk20",
            "level1Cost": "50",
            "level1DisPriority": null,
            "level2Cost": "40",
            "level2DisPriority": null,
            "p2pNegotiationMode": "2_way",
            "p2pPeerIPIgnore": "true",
            "pPPOsicpCheckEnable": "true",
            "silentCost": "false",
            "silentEnable": "false",
            "snpaCheck": "false",
            "typeP2pEnable": "true"
        }
    },
    "existing": {
        "instance": {
            "circuitLevelType": "level_2",
            "ifName": "Eth-Trunk20",
            "level1Cost": "50",
            "level1DisPriority": "10",
            "level2Cost": "40",
            "level2DisPriority": "8",
            "p2pNegotiationMode": null,
            "p2pPeerIPIgnore": "false",
            "pPPOsicpCheckEnable": "false",
            "silentCost": "false",
            "silentEnable": "false",
            "snpaCheck": "false",
            "typeP2pEnable": "false"
        }
    },
    "invocation": {
        "module_args": {
            "bfdblocken": null,
            "bfdstaticen": null,
            "host": "10.190.121.125",
            "ifname": "Eth-Trunk20",
            "instance_id": 3,
            "level1cost": null,
            "level1dispriority": null,
            "level2cost": null,
            "level2dispriority": null,
            "leveltype": null,
            "p2pnegotiationmode": "2_way",
            "p2ppeeripignore": true,
            "ppposicpcheckenable": true,
            "silentcost": null,
            "silentenable": null,
            "snpacheck": null,
            "state": "present",
            "typep2penable": true
        }
    },
    "proposed": {
        "bfdblocken": null,
        "bfdstaticen": null,
        "ifname": "Eth-Trunk20",
        "instance_id": 3,
        "level1cost": null,
        "level1dispriority": null,
        "level2cost": null,
        "level2dispriority": null,
        "leveltype": null,
        "p2pnegotiationmode": "2_way",
        "p2ppeeripignore": true,
        "ppposicpcheckenable": true,
        "silentcost": null,
        "silentenable": null,
        "snpacheck": null,
        "state": "present",
        "typep2penable": true
    },
    "updates": [
        "interface Eth-Trunk20",
        "isis enable 3",
        "isis ppp-negotiation 2-way",
        "isis circuit-type p2p",
        "isis peer-ip-ignore",
        "isis ppp-osicp-check"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************
10.190.121.125             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```
### show on real device
```
[~8850_130.12-Eth-Trunk20]di th
#
interface Eth-Trunk20
 undo portswitch
 ip address 30.2.2.6 255.255.255.0
 isis enable 3
 isis circuit-type p2p
 isis circuit-level level-2
 isis cost 50 level-1
 isis cost 40 level-2
 isis ppp-negotiation 2-way
 isis peer-ip-ignore
 isis ppp-osicp-check
#
return
[~8850_130.12-Eth-Trunk20]

```

### CODE REVIEW

**Comment by xuxiaowei0512** [comment]
-label bug

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 1]
```diff
@@ -0,0 +1,819 @@
...
+                    'supported_by': 'community'}
+
+DOCUMENTATION = """
+---
+module: ce_is_is_interface
+version_added: "2.10"
+author: xuxiaowei0512 (@CloudEngine-Ansible)
+short_description: Manages isis interface configuration on HUAWEI CloudEngine devices.
+description:
+    - Manages  isis process id, creates a isis instance id or deletes a process id
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Manages  isis process id --->  Manages isis process id

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+    type: int
+  ifname:
+    description:
+      -  a L3 interface port.
+    required: true
+    type: str
+  leveltype:
+    description:
+      - level type for three types.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: required: false is default behavior.

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 3]
```diff
@@ -0,0 +1,819 @@
...
+    description:
+      - the dispriority of the level1.
+        The value is a number of 1 to 127.
+    required: false
+    type: int
+  level2dispriority:
+    description:
+      - the dispriority of the level1.
+        The value is a number of 1 to 127.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Same....

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 4]
```diff
@@ -0,0 +1,819 @@
...
+    description:
+      - level type for three types.
+    required: false
+    type: str
+    choices: ['level_1', 'level_2', 'level_1_2']
+  level1dispriority:
+    description:
+      - the dispriority of the level1.
+        The value is a number of 1 to 127.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Same....

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 5]
```diff
@@ -0,0 +1,819 @@
...
+    type: int
+  silentenable:
+    description:
+      - enable the interface can send isis message.
+        The value is a bool type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 6]
```diff
@@ -0,0 +1,819 @@
...
+    type: bool
+  silentcost:
+    description:
+      - Specifies whether the routing cost of the silent interface is 0.
+        The value is a bool type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 7]
```diff
@@ -0,0 +1,819 @@
...
+    type: bool
+  typep2penable:
+    description:
+      - Simulate the network type of the interface as P2P.
+        The value is a bool type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+  snpacheck:
+    description:
+      - Enable SNPA check for LSPs and SNPs.
+        The value is a bool type.
+    required: false
+    type: bool
+  p2pnegotiationmode:
+    description:
+       - Set the P2P neighbor negotiation type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+    description:
+      - Simulate the network type of the interface as P2P.
+        The value is a bool type.
+    required: false
+    type: bool
+  snpacheck:
+    description:
+      - Enable SNPA check for LSPs and SNPs.
+        The value is a bool type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+    description:
+      - When the P2P hello packet is received, no IP address check is performed.
+        The value is a bool type.
+    required: false
+    type: bool
+  ppposicpcheckenable:
+    description:
+      - Interface for setting PPP link protocol to check OSICP negotiation status.
+        The value is a bool type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 11]
```diff
@@ -0,0 +1,819 @@
...
+    description:
+       - Set the P2P neighbor negotiation type.
+    required: false
+    type: str
+    choices: ['2_way', '3_way', '3_wayonly']
+  p2ppeeripignore:
+    description:
+      - When the P2P hello packet is received, no IP address check is performed.
+        The value is a bool type.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 12]
```diff
@@ -0,0 +1,819 @@
...
+    description:
+      - Interface for setting PPP link protocol to check OSICP negotiation status.
+        The value is a bool type.
+    required: false
+    type: bool
+  level1cost:
+    description:
+      - Specifies the link cost of the interface when performing Level-1 SPF calculation.
+        The value is a number of 0 to 16777215.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py [hunk 13]
```diff
@@ -0,0 +1,819 @@
...
+  bfdblocken:
+    description:
+      - Blocking interfaces to dynamically create BFD features.
+        The value is a bool type.
+    required: false
+    type: bool
+  state:
+    description:
+      - Determines whether the config should be present or not on the device.
+    required: false
```
Comment (yuandongxx) [inline lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: same

**Review by yuandongxx — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py (hunk 1 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py (hunk 2 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py (hunk 4 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py (hunk 5 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py (hunk 9 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Done

File: lib/ansible/modules/network/cloudengine/ce_is_is_interface.py (hunk 12 above)
Comment (yuandongx) [reply lib/ansible/modules/network/cloudengine/ce_is_is_interface.py]: Done

**Comment by xuxiaowei0512** [comment]
ready_for_review

**Comment by hwDCN** [comment]
shipit

**Review by yuandongxx — APPROVED** [review]

**Comment by hwDCN** [comment]
shipit

**Comment by hwDCN** [comment]
+1

**Comment by hwDCN** [comment]
shipit

**Pull request merged**